### PR TITLE
Add docs and tests for Requests support

### LINKED ISSUE #66: Document how to use truststore with `requests`
I expect this will be a FAQ

### PR DESCRIPTION
**Pull request by sethmlarson**
Closes #66

### CODE REVIEW

File: docs/source/index.md [hunk 1]
```diff
@@ -28,25 +28,73 @@ The `truststore` module has a single API: `truststore.SSLContext`
    ctx = truststore.SSLContext()
 ```
 
+### Using truststore with pip
+
+Pip v22.2 released with experimental support for verifying certificates with system trust stores with truststore. To enable the feature use the flag `--use-feature=truststore` when installing a package like so:
```
Comment (davisagli) [inline docs/source/index.md:33]:
```suggestion
Pip v22.2 includes experimental support for verifying certificates with system trust stores using `truststore`. To enable the feature, use the flag `--use-feature=truststore` when installing a package like so:
```

File: docs/source/index.md [hunk 2]
```diff
@@ -28,25 +28,73 @@ The `truststore` module has a single API: `truststore.SSLContext`
...
+
+```{code-block} bash
+   # Install Django using system trust stores
+   $ python -m pip install --use-feature=truststore Django
+```
+
+This requires 'truststore' to be installed in the same environment as the one running pip and to be running Python 3.10 or later. For more information you can read the pip documentation about the feature.
```
Comment (davisagli) [inline docs/source/index.md:40]:
```suggestion
This requires `truststore` to be installed in the same environment as the one running pip and to be running Python 3.10 or later. For more information you can read the pip documentation about the feature.
```

Should we link to https://pip.pypa.io/en/stable/user_guide/#using-system-trust-stores-for-verifying-https ?

**Review by davisagli — COMMENTED** [review]
Thanks!

File: docs/source/index.md (hunk 2 above)
Comment (sethmlarson) [reply docs/source/index.md:40]: Oh right, forgot to put a link because of https://github.com/pypa/pip/pull/11322 so wasn't sure if we should wait or not.

Comment (sethmlarson) [reply docs/source/index.md:40]: Looks like there's a redirect being put in for the heading so we should be good to use it anyways. We can update once that PR gets merged.

**Comment by sethmlarson** [comment]
Interesting the Python 3.10 and macOS for Requests seems to be consistently failing on `revoked.badssl.com` but urllib3 works fine. I wonder why that is? :confused:

**Comment by davisagli** [comment]
I'm busy this morning but will take a look later. Maybe the flags related to revocation checks are not getting set correctly when the certifi CA certs have been loaded.

**Comment by davisagli** [comment]
@sethmlarson I added code to explicitly configure macOS security policy to require revocation checks when verify_flags includes VERIFY_CRL_CHECK_CHAIN. I'm still not really sure why the behavior was different when testing with requests; it seems the presence of the `certifi` CAs somehow caused macOS to either not do revocation checks it was otherwise doing, or to ignore failures when trying to do them.

**Review by davisagli — APPROVED** [review]
@sethmlarson approving so it's mergeable since you opened the PR, but I'd appreciate if you could take a look at the part I added first.

File: src/truststore/_macos.py [hunk 3]
```diff
@@ -286,20 +293,37 @@ def _verify_peercerts_impl(
...
+        policies = ssl_policy
+        if ssl_context.verify_flags & ssl.VERIFY_CRL_CHECK_CHAIN:
+            # Add explicit policy requiring positive revocation checks
+            policies = CoreFoundation.CFArrayCreateMutable(
+                CoreFoundation.kCFAllocatorDefault,
+                0,
+                ctypes.byref(CoreFoundation.kCFTypeArrayCallBacks),
+            )
+            CoreFoundation.CFArrayAppendValue(policies, ssl_policy)
+            revocation_policy = Security.SecPolicyCreateRevocation(
```
Comment (sethmlarson) [inline src/truststore/_macos.py:320]: @davisagli Do we need to `CFRelease` the `ssl_policy` and `revocation_policy` objects below as well?

Comment (davisagli) [reply src/truststore/_macos.py:320]: @sethmlarson Yes, having reviewed docs about CoreFoundation memory management more closely, I think you're right.

**Review by davisagli — APPROVED** [review]

File: tests/test_api.py [hunk 4]
```diff
@@ -38,6 +40,8 @@ class FailureHost:
             "Hostname mismatch, certificate is not valid for 'wrong.host.badssl.com'",
             # macOS
             "certificate name does not match",
+            # macOS with revocation checks
+            "certificates do not meet pinning requirements",
```
Comment (sethmlarson) [inline tests/test_api.py:44]: @davisagli This is a strange message for the situation, makes me think of certificate pinning or macOS minimum requirements for certs. What's the full error message here if you have that handy?

Comment (davisagli) [reply tests/test_api.py:44]: `SSLCertVerificationError('"*.badssl.com","R3","ISRG Root X1" certificates do not meet pinning requirements')`

Here is someone else configuring a revocation security policy similar to us, who noted that it's returning `errSecIncompleteCertRevocationCheck` = -67635 indicating that a revocation check could not be completed (which makes more sense), but that CFErrorCopyDescription is returning the wrong error message for that error code: https://developer.apple.com/forums/thread/706629

Comment (sethmlarson) [reply tests/test_api.py:44]: Nice find!! That looks like exactly what we're experiencing.

**Pull request merged**